### ISSUE TITLE
Feature Suggestion: Go back to being original user.

### ISSUE THREAD

**Issue by benmehlman** [stochastic-technologies/django-loginas#29]
Nice job thanks!

You're already storing the original user.pk in the session, it would be great to be able to go back to being who you were without logging out.  If there's no objection I'd be happy to add a view for that.

Then perhaps a context processor to put html for a button in the context when the user is being impersonated.  It would be up to the application developer to decide where to place the button so that anyone will be able to see it, since that would vary for every application.

**Comment by skorokithakis** [comment]
Hmm, this is going to allow privilege escalation if a user somehow manages to store that flag in their session, so I'm not thrilled about it. However, this is useful functionality, so I'd be okay with it if it were optional, off by default and clearly communicated in the documentation that it's potentially insecure.

What do you think?

**Comment by benmehlman** [comment]
How about signing the session information using django.core.signing?  Then it would be as secure as any other django login.. the difficulty of forging the session would be the same as the difficulty of forging the login cookie itself.

But if, even with that, you'd rather have it disabled by default, I'm totally ok with that.

**Comment by skorokithakis** [comment]
That's actually a great idea. Can you make sure to include the original user ID, the target user ID and a timestamp in the signature? It should be something like:

```
request.session[session_flag] = "%s|%s|%s|%s" % (original_user_pk, user.pk, time.time(), signature)
```

And then, to validate:

```
original_id, user_id, timestamp, signature = request.session[session_flag].split("|")
if timestamp < time.time() - 24 * 3600 or user_id != request.user.id: <deny>
```

Something like that would probably provide enough security that there wouldn't be problems. I'd still *prefer* it if this were off by default, since it's really easy to turn it on (and the user would probably have to set up the view anyway), but it's not a requirement.

**Comment by benmehlman** [comment]
Ok, I'm not sure that the time aspect is really needed.. the session already has a timeout that can be configured.  This is an admin function, at some point I feel it is up to the user to take responsibility...

**Comment by skorokithakis** [comment]
The expiration isn't for the data, it's for the signature itself. It's so someone can't steal the signature and replay it at a later time. It's probably overkill, but it's so simple to add that it's probably not worth omitting.

**Comment by benmehlman** [comment]
Hi, sorry I went silent.. things got busy here but I still need this feature and going to get to it in the next week or two.

**Comment by skorokithakis** [comment]
Sure, no problem. I think implementing the feedback above is enough for me to merge and release, thanks!

**Comment by benjaoming** [comment]
Yes, would greatly appreciate if this functionality is added and released :) Oh look, someone added it :)

**Comment by skorokithakis** [comment]
Looks good overall, there are some minor issues I've left comments on. If you can fix those and the failing tests, I can merge, thank you!

**Comment by skorokithakis** [comment]
Merged #31, thanks @benjaoming!
